### DISCUSSION TITLE
Frontend Committer

### DISCUSSION THREAD

**Discussion by LautaroLasorsa** [Consensys/gnark#983] [Q&A]
Hello, I need help to understand which is the function of frontend.Committer

In https://github.com/Consensys/gnark/blob/master/std/recursion/plonk/verifier_test.go there is the following circuit:
```
func (c *InnerCircuitParametric) Define(api frontend.API) error {
	res := c.X
	for i := 0; i < c.parameter; i++ {
		res = api.Mul(res, res)
	}
	api.AssertIsEqual(res, c.Y)

	commitment, err := api.(frontend.Committer).Commit(c.X, res)
	if err != nil {
		return err
	}

	api.AssertIsDifferent(commitment, res)

	return nil
}
```

I don't understand with do I need to use the Commit function and why the check is ```api.AssertIsDifferent``` instead of ```api.AssertIsEqual```. 

Where else do I need to use the committer?

**Comment by ivokub** [comment]
The `Commit()` function computes a vector commitment of the given inputs which can be used as a Fiat-Shamir challenge inside the circuit. The proof verifier can verify that the commitment indeed has been computed correctly. As the commitment computation is specific to the used backend and may not be supported, then instead of exposing it in `frontend.API` directly, we made it implementable and checkable by implicit typing. This is why we have to assert if `frontend.API` implements `frontend.Committer` as it allows to revert to a fallback option in-circuit if the backend doesn't support it.

Now, when we define a circuit and `Commit()` is not used, then the proof is different in case `Commit()` is used (there is additional proof element to check proof of knowledge of the inputs). In the given test we wanted to check that the recursive verifier supports the case where we use the `Commit` function explicitly (it is otherwise a dependency of some packages e.g. field emulation). And we wanted to assert something about the computed commitment to ensure we wouldn't optimize the variable because not being used in further constraint, so added `api.AssertIsDifferent()` call. We cannot easily precompute the commitment as it depends on the proving key and this is generated anew during tests, so cannot easily use `api.AssertIsEqual()`.

By default you do not need to use `frontend.Committer` directly. It is used in the std packages e.g. field emulation, ECDSA signature verification etc. We only have it in tests to avoid any regressions.
